### DISCUSSION TITLE
How to share code between script-scope things?

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#3926] [Q&A] [answered]
[ Slightly simplified version of my problem: ]

I have two targets in the same file that both run some code in an `on_load` method.  Is there a way that I can declare variables that are visible in both of their `on_load` functions?  I've tried declaring them at the global scope in the file, but that's declaration scope and so doesn't work.

**Comment by waruqi** [comment, marked as answer]
### use memcache

```lua
target("foo")
    on_load(function (target)
        import("core.cache.memcache")
        memcache.set("cachename", "key", "value")
    end)

target("bar")
    on_load(function (target)
        import("core.cache.memcache")
        local memcache.get("cachename", "key")
    end)
```

### use target:data()


```lua
target("foo")
    on_load(function (target)
        target:data_set("key", "value")
    end)

target("bar")
    on_load(function (target)
        import("core.project.project")
        local foo = project.targets()["foo"]
        local value = foo:data("key")
    end)
```

### use single script file

```lua
target("foo")
    on_load(function (target)
          import("load_module").load_foo(target)
    end)

target("bar")
    on_load(function (target)
          import("load_module").load_bar(target)
    end)
```

load_module.lua

```lua
function load_foo(target)
    _g.key = "value"
end

function load_bar(target)
    local value = _g.key
end
```

you can also use memcache in this file or use global lua variable

**Reply by davidchisnall** [reply]
Thanks.  The second and third of these are both useful for me in different contexts.

I find that I repeat a lot of `import` directives in functions.  It would be nice if I could do something at the top level in declaration scope in a file to define a set of imports that are added to all script-scope things in the file.

**Reply by davidchisnall** [reply]
I have also been unable to make the `target:set` work with objects, so I ended up having to set my object as a bunch of properties.  I vaguely remember being able to do something like `{expand=false}` as a third argument, but I wasn't able to make that work this time.
